core/src/main/java/org/eclipse/hono/util/EncodedPassword.java

```diff
@@ -0,0 +1,166 @@
+ * Copyright (c) 2018 Red Hat Inc and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat - initial creation
+ *******************************************************************************/
+package org.eclipse.hono.util;
+
+import io.vertx.core.json.JsonObject;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+import java.util.Base64;
+import java.util.Objects;
+
+/**
+ * Helper class for parsing and encoding passwords.
+ * <p>
+ * The encoded password have the following main elements:
+ * <ul>
+ *     <li><em>Hash function</em> that are used to encode the password</li>
+ *     <li><em>Salt</em> which is optionally used for encoding</li>
+ *     <li><em>Password</em> we want to encode in plain text form or hashed value which is used for matching</li>
+ * </ul>
+ * <p>
+ */
+public class EncodedPassword {
+    private static final String PREFIX = "{";
+    private static final String SUFFIX = "}";
+
+    byte[] salt;
+    String password;
+    String hashFunction =  CredentialsConstants.DEFAULT_HASH_FUNCTION;
+
+    /**
+     * Creates a new instance from Hono Secret.
+     *
+     * The secret is expected to be of type <em>hashed-password</em> as defined by
+     * <a href="https://www.eclipse.org/hono/api/credentials-api/#hashed-password">Hono's Credentials API</a>.
+     *
+     * @param honoSecret JSON object that contains hono-formatted secret
+     * @return Encoded password
+     * @throws IllegalArgumentException if candidate hashed-password secret does not contain a password hash
+     */
+    public static EncodedPassword fromHonoSecret(final JsonObject honoSecret) throws IllegalArgumentException {
+        Objects.requireNonNull(honoSecret);
+
+        final String pwdHash = honoSecret.getString(CredentialsConstants.FIELD_SECRETS_PWD_HASH);
+        if (pwdHash == null) {
+            throw new IllegalArgumentException("candidate hashed-password secret does not contain a pwd hash");
+        }
+
+        final String hashFunction = honoSecret.getString(
+                CredentialsConstants.FIELD_SECRETS_HASH_FUNCTION,
+                CredentialsConstants.DEFAULT_HASH_FUNCTION);
+
+        final String encodedSalt = honoSecret.getString(CredentialsConstants.FIELD_SECRETS_SALT);
+        final EncodedPassword encodedPassword = new EncodedPassword();
+        encodedPassword.hashFunction = hashFunction;
+        encodedPassword.password = pwdHash;
+        if (encodedSalt != null) {
+            encodedPassword.salt = Base64.getDecoder().decode(encodedSalt);
+        }
+        return encodedPassword;
+    }
+
+    /**
+     * Creates a new instance from the String in {salt}password format. The password can be in raw or hashed form.
```

Comment:
Does *raw* mean *clear text*? If so, how can it not matter if the password is either the *clear text* password or the password hash?